### DISCUSSION TITLE
Questions about the simple example in the docs

### DISCUSSION THREAD

**Discussion by wigging** [dls-controls/pythonSoftIOC#150] [Q&A] [answered]
I have some questions related to the simple example (shown below) that is given in the [docs](https://dls-controls.github.io/pythonSoftIOC/master/index.html).

```python
# Import the basic framework components.
from softioc import softioc, builder
import cothread

# Set the record prefix
builder.SetDeviceName("MY-DEVICE-PREFIX")

# Create some records
ai = builder.aIn('AI', initial_value=5)
ao = builder.aOut('AO', initial_value=12.45, on_update=lambda v: ai.set(v))

# Boilerplate get the IOC started
builder.LoadDatabase()
softioc.iocInit()

# Start processes required to be run after iocInit
def update():
    while True:
        ai.set(ai.get() + 1)
        cothread.Sleep(1)


cothread.Spawn(update)

# Finally leave the IOC running with an interactive shell.
softioc.interactive_ioc(globals())
```

An `ao` variable is defined but it isn't actually used elsewhere in the code. So why not do something like the snippet shown below instead of assigning it to a variable?
 `_ = builder.aOut('AO', initial_value=12.45, on_update=lambda v: ai.set(v))`

Also, is the use of `globals()` necessary for the `softioc.interactive_ioc(globals())` function? The example seems to work fine if I remove the globals such as `softioc.interactive_ioc()`.

**Comment by AlexanderWells-diamond** [comment, marked as answer]
You are correct regarding `ao`, there is no need to assign it to a variable at all. It's a stylistic choice to keep the lines looking neater together and to indicate to users that this function does also return an object you can interact with.

The use of the `globals()` function is important in making a Python interpreter that functions as one would expect. Try removing it and then inspecting `ai` - you'll find that the interpreter will tell you that it's undefined. By passing the `globals()` dict in here, users can access all their already defined variables in their IOC.

**Reply by wigging** [reply]
I try to avoid defining global variables in Python code and define things in a local scope. The example below is the async example running in a function. Also, instead of using `globals()` or `locals()`, why not just pass a dictionary of variables to the `softioc.interactive_ioc()` function?

```python
from softioc import softioc, builder, asyncio_dispatcher
import asyncio

def main():
    dispatcher = asyncio_dispatcher.AsyncioDispatcher()

    builder.SetDeviceName("MY-DEVICE-PREFIX")
    ai = builder.aIn("AI", initial_value=5)
    ao = builder.aOut("AO", initial_value=12.45, always_update=True, on_update=lambda v: ai.set(v))

    builder.LoadDatabase()
    softioc.iocInit(dispatcher)

    async def update():
        while True:
            ai.set(ai.get() + 1)
            await asyncio.sleep(1)

    dispatcher(update)
    softioc.interactive_ioc({"ai": ai, "ao": ao})

if __name__ == "__main__":
    main()
```

**Reply by AlexanderWells-diamond** [reply]
The examples are there to show a minimal way to run an IOC. Your code is obviously preferred for an actual program, but these examples are not expected to be taken and used as-is. Your example provides no additional insight into how pythonSoftIOC behaves.

You are again correct that you could manually define the dictionary list to pass to `interactive_ioc`. However, this will become increasingly hard to maintain as more and more variables (and functions) are defined across the program. If you wish to use it like that you are most welcome to, but I would not advise doing it that way.
